_posts/2016-12-12-Reactive-Programming-Spring-Reactor.md

```diff
@@ -0,0 +1,525 @@
+Some of the method names will be different to RxJava2, but the logic behind these methods has been aligned among RxJava and Spring .
+
+
+```
+Flux.just(“red”, “white”, “blue”)
+       .flatMap(carRepository::findByColor)
+       .collect(Result:: new, Result::add)
+       .doOnNext(Result::stop)
+       .subscribe(doWithResult);
+
+
+Interface CarRepository {
+    Flux<Car> findByColor(String color);
+}
+
+```
+This Flux will retrieve all cars which match the color "red" then those with the color "white" and finally "blue".
+So instead of just 3 elements, after this Flatmap we are going to have a lot more elements.
+This is all handled with backpressure in mind, fore example when the flatmap is busy merging data we will not ask for extra records
+
+If the Repository implements Flux as a method signature, it will be picked up automatically as a reactive repository.
+This support for Flux will be part of the whole of Spring 5.
+Spring Data, Spring Security, Spring MVC, ... are all good candidates who will have this kind of support.
+
+## Mono
+<p style="text-align: center;">
+  <img alt="a Mono" src="/img/reactive/mono.png">
+</p>
+
+None is like a flux, but will return at most 1 result, so it does have less methods.
+
+```
+Mono.delayMillis(3000)
+    .map(d -> "Spring 4"°
+    .or(Mono.delayMillis(2000).map(d -> "Spring 5"))
+    .then(t -> Mono.just(t + " world"))
+    .elapsed()
+    .subscribe()
+
+
+```
+This Mono will wait for 3 seconds on the "call" to Spring 4 or 2 seconds on that of Spring 5.
+The fastest result will be the one which will be outputted.
+
+
+The Mono has as advantage over an Observable Future of Java 8 that a Mono will only be triggered if you subscribe to it.
+While with an Observable the call to send() will execute the operation.
+
+
+
+<a name="testing"/>
+
+# Testing
+
+Block() exists for very specific use cases and for testing.
+Never, ever use this in production, as is it blocks your call, which does infer with the Reactive non-blocking statements. ;-)
+
+```
+Mono.delayMillis(3000)
+    .map(d -> "Spring 4"°
+    .or(Mono.delayMillis(2000).map(d -> "Spring 5"))
+    .then(t -> Mono.just(t + " world"))
+    .elapsed()
+    .block()
+
+
+```
+
+You can also make use of [Stepverifier](ttp://next.projectreactor.io/ext/docs/api/reactor/test/StepVerifier.html) to test Flux, Mono and any other kind of Reactive Streams Publisher.
+
+```
+
+@Test
+public void expectElementsWithThenComplete() {
+    expectSkylerJesseComplete(Flux.just(new User("swhite", null, null), new User("jpinkman", null, null)));
+}
+
+// Use StepVerifier to check that the flux parameter emits a User with "swhite" username and another one with "jpinkman" then completes successfully.
```

Comment:
Maybe 2 code blocks with text in between and no comment?
Pure preference probably.